client/asset/dcr/dcr.go

```diff
@@ -0,0 +1,1165 @@
+			}
+			if fee == reqFee || (reqFee < fee && tried[reqFee] == 1) {
+				// If we've already tried it, we're on our way back up, so this fee is
+				// likely as good as it gets.
+				break
+			}
+			// We must have some room for improvement
+			tried[fee] = 1
+			fee = reqFee
+			changeOutput.Value = int64(remaining - fee)
+			continue
+		}
+	}
+
+	checkHash := msgTx.TxHash()
+	dcr.log.Debugf("%d signature cycles to converge on fees for tx %s", sigCycles, checkHash)
+	txHash, err := dcr.node.SendRawTransaction(msgTx, false)
+	if err != nil {
+		return nil, err
+	}
+	if *txHash != checkHash {
+		return nil, fmt.Errorf("transaction sent, but received unexpected transaction ID back from RPC server. "+
+			"expected %s, got %s", *txHash, checkHash)
+	}
+	if !isDust {
+		dcr.addChange(txHash, uint32(changeIdx))
+	}
+	return msgTx, nil
+}
+
+// For certain dcrutil.Address types.
+type signatureTyper interface {
+	DSA() dcrec.SignatureType
+}
+
+// createSig creates and returns the serialized raw signature and compressed
+// pubkey for a transaction input signature.
+func (dcr *ExchangeWallet) createSig(tx *wire.MsgTx, idx int, pkScript []byte, addr dcrutil.Address) (sig, pubkey []byte, err error) {
+	sigTyper, ok := addr.(signatureTyper)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid address type")
+	}
+
+	priv, pub, err := dcr.getKeys(addr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sigType := sigTyper.DSA()
+	switch sigType {
+	case dcrec.STEcdsaSecp256k1:
+		sig, err = txscript.RawTxInSignature(tx, idx, pkScript, txscript.SigHashAll, priv)
+	default:
+		sig, err = txscript.RawTxInSignatureAlt(tx, idx, pkScript, txscript.SigHashAll, priv, sigType)
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return sig, pub.SerializeCompressed(), nil
+}
+
+// getKeys fetches the private/public key pair for the specified address.
+func (dcr *ExchangeWallet) getKeys(addr dcrutil.Address) (*secp256k1.PrivateKey, *secp256k1.PublicKey, error) {
+	wif, err := dcr.node.DumpPrivKey(addr, chainParams.PrivateKeyID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	priv, pub := secp256k1.PrivKeyFromBytes(wif.PrivKey.Serialize())
+	return priv, pub, nil
+}
+
+// monitorConnection monitors the dcrwallet connection, sending an error via the
+// tipChange function when a disconnect is detected.
+func (dcr *ExchangeWallet) monitorConnection() {
+	ticker := time.NewTicker(time.Second * 10)
+out:
+	for {
+		select {
+		case <-ticker.C:
+			if dcr.node.Disconnected() {
+				dcr.tipChange(fmt.Errorf("node disconnected"))
+			}
+		case <-dcr.ctx.Done():
+			break out
+		}
+	}
+}
+
+// Convert the DCR value to atoms.
+func toAtoms(v float64) uint64 {
+	return uint64(math.Round(v * 1e8))
+}
+
+// toCoinID converts the tx hash and vout to a coin ID, as a []byte.
+func toCoinID(txHash *chainhash.Hash, vout uint32) []byte {
+	coinID := make([]byte, 36)
+	copy(coinID[:32], txHash[:])
+	b := make([]byte, 4)
+	binary.BigEndian.PutUint32(b, vout)
+	copy(coinID[32:], b)
```

Comment:
Similar optimization as in the BTC wallet PR (`PutUint32` directly into `coinID`).